### DISCUSSION TITLE
lifespan context is not stored in state, it actually replaces it

### DISCUSSION THREAD

**Discussion by SirDroun** [encode/starlette#2515] [Potential Issue] [answered]
Using the lifespan pattern with a context manager and am quite happy with its existence. There is still something that bothers me with with it though. The context that the context manager creates needs to be None, because anything else is regarded as the full `state: State()` of the the app. It is possible to set this manually in the state, but that does not fit well with a context created by some other object.

Example:
```
class MyState:
  def __init__(self, application):
    self.some_connecion = connect()

  @asynccontextmanager
  @classmethod
  async def lifespan(cls, application):
    yield cls(application)

app = Starlette(..., lifespan=MyState.lifespan)

def some_endpoint(...):
    request.app.state.lifespan.some_connection.do_stuff()
```

This will fail today, `application.state` becoming `MyState()` and this is not what is expected by a number of other functions.
You would need to do:
```
 async def lifespan(cls, application):
   my_state = MyState(application)
   application.state.lifespan = my_state
   yield
```
Which i find counter intuitive for a context manager.

I would suggest changing in `routing.Router.lifespan()` to check if the returned context is a `State()` instance (to keep current behavior) and if not, add the context as a `lifespan` part of the state, or add it as a separate `local_state` or similar on the application level.

**Comment by Kludex** [comment, marked as answer]
You need to use `request.state` in the endpoint, not `app.state`.

**Reply by SirDroun** [reply]
Sorry, but I don't think that is the issue. Yes, my example was wrong, but the fact it is that the lifespan code is not handling this. In https://github.com/encode/starlette/blob/master/starlette/routing.py#L725 you have
```
            async with self.lifespan_context(app) as maybe_state:
                if maybe_state is not None:
                    if "state" not in scope:
                        raise RuntimeError(
                            'The server does not support "state" in the lifespan scope.'
                        )
                    scope["state"].update(maybe_state)
```
And if I'm not mistaken the `scope["state"].update(maybe_state)` will replace whatever is in "state", i.e removing everything else that that has previously been assigned to the state. That is also the symptom I see when trying to return other than None from the lifespan ctx manager.

**Reply by Kludex** [reply]
There was nothing assigned to it at that point. Just an empty dict.
